sdk/nodejs/query/interfaces.ts

```diff
@@ -0,0 +1,44 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// NOTE: We don't need `Disposable` or `Enumerator#reset()` yet, but it's worth noting here that we
+// didn't include them, and this causes us to diverge slightly from the "canonical" `Enumerable`
+// model.
+//
+
+export interface Enumerable<T> {
+    filter(f: (t: T) => boolean): Enumerable<T>;
+    flatMap(f: (t: T, index?: number) => T[]): Enumerable<T>;
+    map<U>(f: (t: T) => U): Enumerable<U>;
+    take(n: number): Enumerable<T>;
+
+    toArray(): Promise<T[]>;
```

Comment:
If we do end up sticking with these interfaces... Is it by design that this returns `Promise<T[]>` instead of `T[]`? None of the other functions on this interface return promises.

Reply:
This is on purpose. Once you start to do stuff async, you can't "get out of" the `Promise`. The way you would use this is:

```
const myarr = foo.filter(...).toArray();
```